src/Farmer/Builders/Builders.Dns.fs

```diff
@@ -0,0 +1,335 @@
+
+    /// Sets the target resource of the record.
+    [<CustomOperation "target_resource">]
+    member _.RecordTargetResource(state:AaaaRecordProperties, targetResource) = { state with TargetResource = Some targetResource }
+
+type DnsNsRecordBuilder() =
+    member __.Yield _ = { NsRecordProperties.NsdNames = []; Name = ResourceName.Empty; TTL = None; }
+    member __.Run(state : NsRecordProperties) : DnsZoneRecordConfig =
+        { emptyRecord with
+            Name =
+                if state.Name = ResourceName.Empty then ResourceName "@"
+                else state.Name
+            TTL =
+                if state.TTL = None then failwith "You must set a TTL"
+                else state.TTL.Value
+            NsRecords = state.NsdNames
+            Type = Some NS }
+
+    /// Sets the name of the record set.
+    [<CustomOperation "name">]
+    member _.RecordName(state:CNameRecordProperties, name) = { state with Name = name }
+    member this.RecordName(state:CNameRecordProperties, name:string) = this.RecordName(state, ResourceName name)
+
+    /// Add NSD names
+    [<CustomOperation "add_nsd_names">]
+    member _.RecordNsdNames(state:NsRecordProperties, nsdNames) = { state with NsdNames = state.NsdNames @ nsdNames }
+
+    /// Sets the TTL of the record.
+    [<CustomOperation "ttl">]
+    member _.RecordTTL(state:NsRecordProperties, ttl) = { state with TTL = Some ttl }
+
+type DnsPtrRecordBuilder() =
+    member __.Yield _ = { PtrRecordProperties.PtrdNames = []; Name = ResourceName.Empty; TTL = None; }
+    member __.Run(state : PtrRecordProperties) : DnsZoneRecordConfig =
+        { emptyRecord with
+            Name =
+                if state.Name = ResourceName.Empty then ResourceName "@"
+                else state.Name
+            TTL =
+                if state.TTL = None then failwith "You must set a TTL"
+                else state.TTL.Value
+            PtrRecords = state.PtrdNames
+            Type = Some PTR }
+
+    /// Sets the name of the record set.
+    [<CustomOperation "name">]
+    member _.RecordName(state:CNameRecordProperties, name) = { state with Name = name }
+    member this.RecordName(state:CNameRecordProperties, name:string) = this.RecordName(state, ResourceName name)
+
+    /// Add PTR names
+    [<CustomOperation "add_ptrd_names">]
+    member _.RecordPtrdNames(state:PtrRecordProperties, ptrdNames) = { state with PtrdNames = state.PtrdNames @ ptrdNames }
+
+    /// Sets the TTL of the record.
+    [<CustomOperation "ttl">]
+    member _.RecordTTL(state:PtrRecordProperties, ttl) = { state with TTL = Some ttl }
+
+type DnsTxtRecordBuilder() =
+    member __.Yield _ = { TxtRecordProperties.Name = ResourceName.Empty; TxtValues = []; TTL = None; }
+    member __.Run(state : TxtRecordProperties) : DnsZoneRecordConfig =
+        { emptyRecord with
+            Name =
+                if state.Name = ResourceName.Empty then ResourceName "@"
```

Comment:
Throw an exception here, rather than implicitly putting in a default value silently. Or, consider putting this into the default value within `Yield`.

Reply:
Will move it to `Yield`. Went with a silent default as it's what the Portal does as well and is common with these record types.

Reply:
Done

Reply:
Great. If that's what the portal does, it's perhaps not so bad except that it's a "silent" change. I've tried to avoid that with other names as it can cause unintended surprises for authors. However this one seems somewhat more harmless.

Reply:
I'm fine with either cases, a default to "@" or it being required.